Move "installed" and "installed_upstream" from PackageBase to Spec

alalazo opened the pull request:
fixes #30187

This PR moves the `installed` and `installed_upstream` properties from `PackageBase` to `Spec` and is a step towards being able to reuse specs for which we don't have a `package.py` available. It _should_ be sufficient to complete the concretization step and see the spec in the concretized DAG.

To fully reuse a spec without a package.py though we need a way to serialize enough data to reconstruct the results of calls to:
- `Spec.libs`, `Spec.headers` and `Spec.ommand`
- `Package.setup_dependent_*_environment` and `Package.setup_run_environment`
- Other methods ?

This came out of a discussion with @becker33 and @scheibelp during one of the weekly [telco](https://github.com/spack/spack/wiki/Telcon%3A-2022-04-20)

alalazo committed:
Move "installed_upstream" from PackageBase to Spec

alalazo committed:
Move "installed" from PackageBase to Spec

alalazo committed:
Fix two failing unit tests

alalazo committed:
Add a unit-test

alalazo committed:
Fix another regression in unit tests

tgamblin commented on lib/spack/spack/cmd/fetch.py:
```diff
@@ -70,13 +70,9 @@ def fetch(parser, args):
-        package = spack.repo.get(spec)
-        package.do_fetch()
+                s.package.do_fetch()
+        spec.package.do_fetch()
```
Isn’t this redundant if args.missing or args.dependencies is true? Should the traversal above have `root=False`?

alalazo replied:
Fixed in f862dce

tgamblin commented on lib/spack/spack/package.py:
```diff
@@ -1406,22 +1397,6 @@ def virtuals_provided(self):
-        except KeyError:
-            # If the spec is not in the DB, the method
-            #  above raises a Key error
-            return False
```
I suspect a lot of downstream code (spackdev, spack-manager, scripts, etc.) uses these methods, or at least `installed()`. We should add stubs for both of these that work by delegating to the spec, but print deprecation warnings, and remove them in 0.19.

I wonder if it’s worth having some kind of test decorator so we could mark the tests for these to fail after a certain version to remind us to remove them.

alalazo replied:
Fixed in f862dce

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1554,6 +1555,29 @@ def spliced(self):
+        Returns:
+            True if the package has been installed, False otherwise.
+        """
+        try:
```
These methods should either assert that the spec isn’t concrete or just return false if it’s concrete.

I can’t decide which is better. I lean towards asserting since I think it’s probably an error if you ask an abstract spec whether it’s installed, but I could see returning False as maybe a less prickly API. I still kind of think something’s wrong with the client code if it has to ask this of an abstract spec.

alalazo replied:
Fixed in f862dce

tgamblin reviewed (changes requested):
Some minor requests but looks great to me. 

The refactored test where we don’t have  to make a mock package anymore really shows why this is the right refactor — we shouldn’t have to go through Package when we can just ask the DB.

alalazo committed:
Address review comments

- Add stub methods to packages with warnings
- Add a missing "root=False" in cmd/fetch.py
- Assert that a spec is concrete before checking
  installation status
